R/qenv-eval_code.R

```diff
@@ -28,18 +28,20 @@
 setGeneric("eval_code", function(object, code) standardGeneric("eval_code"))
 
 setMethod("eval_code", signature = c("qenv", "character"), function(object, code) {
-  id <- sample.int(.Machine$integer.max, size = 1)
+  parsed_code <- parse(text = code, keep.source = TRUE)
+  comments <- extract_comments(parsed_code)
+  id <- sample.int(.Machine$integer.max, size = length(parsed_code))
 
   object@id <- c(object@id, id)
   object@env <- rlang::env_clone(object@env, parent = parent.env(.GlobalEnv))
-  code <- paste(code, collapse = "\n")
-  object@code <- c(object@code, code)
+  object@code <- c(object@code, trimws(paste(as.character(parsed_code), comments)))
 
-  current_warnings <- ""
-  current_messages <- ""
+  current_warnings <- rep("", length(parsed_code))
+  current_messages <- rep("", length(parsed_code))
 
-  parsed_code <- parse(text = code, keep.source = TRUE)
-  for (single_call in parsed_code) {
+
+  for (i in seq_along(parsed_code)) {
+    single_call <- parsed_code[i]
```

Comment:
Parsing independently calls and comments is a risky business. Also we distort original `code` which is important for a [future issue](https://github.com/insightsengineering/teal.code/issues/212).

```suggestion
  code_split <- split_code(code) # keep original character
  for (i in seq_along(code_split)) {
    single_call <- code_split[[i]]
    ...
```

Reply:
Also co-ocurrence and side-effects can be extracted when evaluating attach this information to each call.
Then `get_code_dependencies` will be simplified

Reply:
`split_code` could split into calls, but @code needs to be stored with comments. If you use parse and `utils::getParsedData` then you'll need to convert back to the character. I think using `getParsedData` will be the most accurate to determine the end of the call.

Reply:
I think we can work on a simplified scenario. Let's just limit ourself to the current input and functions.

```r
code <- "
  # initial comment
  a <- 1 # A comment
  b <- 2 # inline comment
  c <- 3 # C comment
  # inbetween comment
  d <- 4
  # finishing comment
"

parsed_code <- parse(text = code)
comments <- extract_comments(parsed_code)

for(single_call in parsed_code) eval(single_call)

trimws(paste(as.character(parsed_code), comments))
```

Pay attention that
- `eval` inside `for` loop works on `expressions`
- we do not have a way to divide `code` into a calls in another way than through the inbuild parser or `parse(text = `
- we are able to `extract_comments` in our way, even thought sometimes there is more comments than calls

> # keep original character

I don't see a way how we can apply `eval` on `character` and how to divide character `code` into calls without parsing.